File: analysis_utils/matlab/load_sc_data.m
function [ data, count, t ] = load_sc_data (varargin)
% load_sc_data Code to load up data which has been output from SpineCreator.
% Returns the time axis in milliseconds in the variable t.

% For Octave, and XML support (using load_sc_data with one argument),
% you need these jar files from e.g.
% https://www.apache.org/dist/xerces/j/Xerces-J-bin.2.11.0.tar.gz
%
% you need use javaaddpath to include these files:
% javaaddpath('/home/you/Downloads/xerces-2_11_0/xercesImpl.jar');
% javaaddpath('/home/you/Downloads/xerces-2_11_0/xml-apis.jar');
%
% Further down in your copy of this code, change the javaaddpath calls
% and then change the gotXerces variable to 1.
    
% varargin can be filled with arguments file_path then optionally
% num_neurons (which avoids use of xml reading code).

    isOctave = exist('OCTAVE_VERSION', 'builtin') ~= 0;

    if nargin == 1
        % No n_neurons, so only extract file_path
        file_path = varargin{1};
        num_neurons = [];
    elseif nargin == 2
        file_path = varargin{1};
        num_neurons = varargin{2};
    else
        display ('Error: Wrong number of arguments');
        return;
    end

    % If file path ends with .bin, or .xml then find the root name.
    bin_end = file_path (end-3:end);
    xml_end = file_path (end-6:end);
    if bin_end == '.bin'
        base_path = file_path (1:end-4);
    elseif xml_end == 'rep.xml'
        base_path = file_path (1:end-7);
    else
        % Error
        display (['Error: Bad SpineCreator log file name: ', ...
                  file_path]);
        return;
    end

    xml_file = [base_path 'rep.xml'];
    bin_file = [base_path '.bin'];

    % Timestep size. Initialise to 0.
    dt = 0;

    % Init some parameters which may be read from the XML file
    logFileType = 'binary';
    logEndTime = 0;
    logType = 'double';
    logPort = '';
    % Obtained from logCol elements:
    logColHeadings = [];
    logColTypes = [];
    
    if (isempty(num_neurons))
        % Find the number of neurons in the binary log file from
        % the xml file.
        if isOctave
            % User! You're using octave and you have asked for XML reading of the
            % SpineML log metadata, so need to javaaddpath.
            gotXerces = 1; % User! Change this to 1 when you're done!
            if gotXerces == 0
                display (['Calling this function with 1 arguments means ' ...
                          'it needs to read XML. For XML support  you need ' ...
                          'to get Xerces and modify the javaaddpath() ' ...
                          'lines in the code. Please see the ' ...
                          'load_sc_data.m code.']);
                return;
            end
            % User! Modify these javaaddpath lines to match the Xerces you downloaded!
            javaaddpath('/home/seb/ownCloud/octave/xerces-2_11_0/xercesImpl.jar');
            javaaddpath('/home/seb/ownCloud/octave/xerces-2_11_0/xml-apis.jar');


            % these 3 lines are equivalent to infoDoc = xmlread (xml_file)
            parser = javaObject ('org.apache.xerces.parsers.DOMParser');
            parser.parse (xml_file);
            infoDoc = parser.getDocument;
        else
            infoDoc = xmlread (xml_file);
        end

        % Assume Analog Log here. Probably wrong for event log.
        logFileType = char(infoDoc.getElementsByTagName ...
                           ('LogFileType').item(0).getFirstChild ...
                           .getData);
        display (logFileType)

        if strcmp(logFileType, 'csv') || strcmp(logFileType, 'binary')
            % ok
        else
            display (['File described by ', xml_path, ' is not marked ' ...
                      'as being in binary or csv format.']);
            return;
        end

        % Log end is in steps of size dt. Unused at present even though
        % this is in the UI? Also may need logStartTime in future to
        % generate t.
        logEndTime = str2num(char(infoDoc.getElementsByTagName ...
                                  ('LogEndTime').item(0).getFirstChild.getData));
        num_neurons = str2num(char(infoDoc.getElementsByTagName ...
                                   ('LogAll').item(0).getAttribute('size')));

        logType = char(infoDoc.getElementsByTagName ...
                       ('LogAll').item(0).getAttribute('type'));
        
        %logHeading = char(infoDoc.getElementsByTagName ...
        %                  ('LogCol').item(0).getAttribute('heading'))
        
        % Timestep is specified in milliseconds
        dt = str2num(char(infoDoc.getElementsByTagName ...
                          ('TimeStep').item(0).getAttribute('dt')));    

    end % else num_neurons already set.

    if strcmp(logFileType, 'binary') % The log file is binary format.

        % First, open the file:
        [ fid, fopen_msg ] = fopen (bin_file, 'r', 'native');
        if fid == -1
            display (['Failed to open file ', file_path, ' with error: ', ...
                      fopen_msg]);
            count = 0; data = [];
            return;
        end

        % Imagine num_neurons is 4. This will return a 4 row matrix with as
        % many columns as there are timesteps in your output time
        % series. SpineCreator data is always double precision.
        [ data, count ] = fread (fid, [num_neurons, Inf], 'double=>double');

        % Construct time series in milliseconds.
        t = [0 : dt : (dt*count)-dt];

        % Finally, close the file.
        %rtn = fclose (fid);
        rtn = 0;
        if rtn == -1
            display (['Warning: failed to close file ', file_path]);
        end
        
    else % The log file is in csv format
        
        display ('Warning, csv file type not yet supported.');
    
    end
end

File: analysis_utils/python/load_sc_data.py
def load_sc_data (file_path, num_neurons=0):
# load_sc_data Code to load up data which has been output from SpineCreator.
# Returns the time axis in milliseconds in the variable t.
#
# file_path is the path to the SpineCreator log file; it can be either
# the .bin file or the .xml file. If it's the .bin file, then
# num_neurons must be provided.

    import array as ar
    import numpy as np
    import struct

    # If file path ends with .bin, or .xml then find the root name.
    end = len(file_path)
    bin_end = file_path [end-4:end]
    xml_end = file_path [end-7:end]
    if (bin_end == '.bin'):
        base_path = file_path [:end-4]
    elif (xml_end == 'rep.xml'):
        base_path = file_path [:end-7]
    else:
        # Error
        print 'Error: Bad SpineCreator log file name: ' + file_path
        return (0,0,0)

    xml_file = base_path + 'rep.xml'
    bin_file = base_path + '.bin'

    # Timestep size. Initialise to 1
    dt = 1

    # Re-implement me!
    if (num_neurons == 0):
        # Find the number of neurons in the binary log file from the
        # xml file.
        import xml.etree.ElementTree as et
        tree = et.parse(xml_file)
        root = tree.getroot()

        # Assumed Analog Log here. May be wrong for event log.
        logFileType = root.find('.//LogFileType')
        if logFileType.text != 'binary':
            print 'File described by ', xml_file, ' is not marked as being in binary format.'
            return(0,0,0)

        # Log end is in steps of size dt. Unused at present even
        # though this is in the UI? Also may need logStartTime in
        # future to generate t.
        logEndTime = float(root.find('.//LogEndTime').text)
        num_neurons = int(root.find('.//LogAll').get('size'))
        # Timestep is specified in milliseconds
        dt = float(root.find('.//TimeStep').get('dt'))
        #print 'End time =', logEndTime, 'ms  num_neurons =', num_neurons, ' dt =', dt, 'ms'

    struct_fmt = 'd'
    struct_len = struct.calcsize(struct_fmt)
    struct_unpack = struct.Struct(struct_fmt).unpack_from

    data = []
    with open(bin_file, "rb") as fid:
        while True:
            dat = fid.read(struct_len)
            if not dat: break
            s = struct_unpack(dat)
            data.append(s[0])
    count = len(data)

    # Construct time series in milliseconds.
    t = np.linspace (0, (dt*count)-dt, count)

    # return data, count, t in a list. Could be a tuple?
    return (data, count, t)
